### ISSUE TITLE
Handling of flyway migrations

### ISSUE THREAD

**Issue by dnlkoch** [terrestris/shogun#263]
If we introduce new flyway migrations (that should for example be necessary for #247 and #258) this will work for a standalone shogun-boot setup, but project implementations might fail during applying the migrations as they might have added migrations already. As an example: Adding `V1.0.0__my_new_shogun_migration.sql` would breakup the migration order if the custom project already contains `V2.0.0__my_existing_project_migration.sql` (`V1.0.0` < `V2.0.0`).

Any ideas on how to get around this problem?

**Comment by weskamm** [comment]
maybe i got it wrong but cant we just have shogun migrations in folder 1 and project migrations in folder 2. Applying always folder 1 before folder 2 should then fix this issue?

**Comment by dnlkoch** [comment]
Unfortunately this is not possible out of the box. Flyway prioritizes the order of the migrations - given by the version in the name of the migration file - over the order of the migrations directories found on the classpath.

I see the following solutions in here:

1. Applying multiple flyway lifecycles.

Currently there is a single flyway lifecycle available only. This applies the given migrations in the logical order and records the migration history in a single table (usually `flyway_schema_history`). Each project could potentially implement a separate lifecycle (by implementing a `FlywayMigrationStrategy`) that applies against a separate history table and is configured to execute migration after the default one always.

Potentially this might be reasonable, but would require several adjustments in projects.

2. Introducing a version convention for shogun and projects.

Flyway supports the config option [outOfOrder](https://flywaydb.org/documentation/configuration/parameters/outOfOrder) which disables the strict order while applying the migrations. Having this enabled, projects would able to apply newer SHOGun migrations while they were ahead of the version range available in SHOGun. To avoid any potential conflicts, a version convention must be applied, e.g. having all shogun migrations in version range `0` and projects in range `>0`, e.g.

SHOGun:

```
├── db
    └── migration
        ├── V0.1.0__Baseline.sql
        ├── V0.1.1__Init_permissions.sql
        ├── V0.2.0__Add_Topics.sql
        ├── V0.3.0__Drop_ImageFile_width_and_height_NOT_NULL.sql
        ├── V0.4.0__Add_audit_tables.sql
        ├── V0.5.0__Add_missing_permission_READ_UPDATE_DELETE.sql
        ├── V0.6.0__Remove_topic.sql
        └── V0.x.x__n.sql
```

MyProject:

```
├── db
    └── migration
        ├── V1.0.0__Add_project_entity.sql
        ├── V1.x.x__Init_permissions.sql
        └── V>=1.x.x__n.sql
```

Projects should then just update their `application.yaml`:

```yaml
  flyway:
    (…)
    outOfOrder: true
```

Personally I would go for option 2.

**Comment by marcjansen** [comment]
> Personally I would go for option 2

Same here. 👍

**Comment by dnlkoch** [comment]
In #261 a test has been added to enforce the correct naming convention.

**Comment by dnlkoch** [comment]
Fixed via #285.
